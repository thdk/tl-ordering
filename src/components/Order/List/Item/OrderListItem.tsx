import * as React from "react";

import { connect } from "react-redux";
import { IState } from "../../../../core/app/types";
import { IOrder } from "../../../../core/orders/types";
import OrderLink from "../../../links/OrderLink";

interface IOrderListItemProps {
    order: IOrder;
}

interface IPropsFromState { total: number; }

type Props = IOrderListItemProps & IPropsFromState;

const OrderListItem = (props: Props) => {
    const { order: { id, customerId }, total } = props;

    return (
        <div className="row">
            <div className="row-left">
                <div>{id}</div>
                <div>{customerId}</div>
            </div>
            <div className="row-right">
                <div>{total.toFixed(2)}</div>
                <div>
                    <OrderLink orderId={id}>detail</OrderLink>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state: IState, ownProps: IOrderListItemProps) => {

    const total = state.orderItems.byOrderId[ownProps.order.id].reduce((p, c) => {
        const product = state.products.byId[c.productId];
        if (!product) { console.error("No product found for " + c.productId + ". Using '0' as price"); }
        p += (product ? product.price : 0) * c.quantity;
        return p;
    }, 0);

    return { total };
};

const ConnectedOrderListItem = connect(
    mapStateToProps,
)(OrderListItem);

export default ConnectedOrderListItem;
