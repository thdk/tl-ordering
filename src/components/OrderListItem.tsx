import * as React from 'react';

import { IOrder } from "../interfaces/orders";
import OrderLink from './links/OrderLink';
import { IState } from '../interfaces/state';
import { connect } from 'react-redux';

export interface IOrderListItemProps {
    order: IOrder;
}

export type PropsFromState = { total: number };

type Props = IOrderListItemProps & PropsFromState;

const OrderListItem = (props: Props) => {
    const { order: { id, customerId }, total } = props;

    return (
        <div className="row">
            <div className="row-left">
                <div>{id}</div>
                <div>{customerId}</div>
            </div>
            <div className="row-right">
                <div>{total.toFixed(2)}</div>
                <div>
                    <OrderLink orderId={id}>detail</OrderLink>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state: IState, ownProps: IOrderListItemProps) => {

    const total = state.orderItems.byOrderId[ownProps.order.id].reduce((p, c) => {
        const product = state.products.byId[c.productId];
        if (!product) console.error("No product found for " + c.productId + ". Using '0' as price");
        p += (product ? product.unitPrice : 0) * c.quantity;
        return p;
    }, 0);

    return { total };
};

const ConnectedOrderListItem = connect(
    mapStateToProps
)(OrderListItem);

export default ConnectedOrderListItem;