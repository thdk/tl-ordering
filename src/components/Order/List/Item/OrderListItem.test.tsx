import { shallow, ShallowWrapper } from "enzyme";
import React from "react";
import OrderListItem from "./OrderListItem";

describe("Render OrderListItem", () => {
    let component: ShallowWrapper;

    beforeEach(() => {
        component = shallow(
            <OrderListItem
                order={{
                    id: "o1",
                    total: 15.19,
                    customerId: "c1",
                }}
            />,
        );
    });

    it("should match snapshot", () => {
        expect(component.debug()).toMatchSnapshot();
    });
});
