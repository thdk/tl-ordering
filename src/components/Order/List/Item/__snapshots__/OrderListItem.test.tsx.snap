// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render OrderListItem should match snapshot 1`] = `
"<div className=\\"row\\">
  <div className=\\"row-left\\">
    <div>
      o1
    </div>
    <div>
      c1
    </div>
  </div>
  <div className=\\"row-right\\">
    <div>
      15.19
    </div>
    <div>
      <Component orderId=\\"o1\\">
        detail
      </Component>
    </div>
  </div>
</div>"
`;
